ic-factory/src/state.rs

```diff
@@ -487,45 +495,74 @@ fn consume_message_cycles() -> Result<u64, FactoryError> {
     Ok(ic_kit::ic::msg_cycles_accept(amount))
 }
 
-async fn consume_provided_icp(
-    caller: Principal,
+/// Transfers the ICP from the caller to the factory canister.
+///  We transfer the minimum amount of ICP required to cover the canister creation fee,
+/// and then we top up the Factory canister with the cycles.
+/// We send the remaining ICP to the `icp_to` Principal.
+async fn transfer_and_top_up(
+    icp_fee: u64,
     ledger: Principal,
+    caller: Principal,
     icp_to: Principal,
-    icp_fee: u64,
-) -> Result<(), FactoryError> {
+) -> Result<u64, FactoryError> {
     let id = ic_kit::ic::id();
     let balance = ledger
         .get_balance(id, Some((&PrincipalId(caller)).into()))
         .await
         .map_err(FactoryError::LedgerError)?;
 
-    if balance < icp_fee + DEFAULT_TRANSFER_FEE.get_e8s() {
-        return Err(FactoryError::NotEnoughIcp(
-            balance,
-            icp_fee + DEFAULT_TRANSFER_FEE.get_e8s(),
-        ));
+    // The TOTAL_FEE is 4 times, because of the different transfers :-
+    // 1. Transfer from caller to factory
+    // 2. Top up factory (`send_dfx` and `notify_dfx` calls)
+    // 3. Transfer remaining ICP from factory to `icp_to` Principal
+    const TOTAL_FEE: u64 = DEFAULT_TRANSFER_FEE.get_e8s() * 4;
+
+    if balance < icp_fee + TOTAL_FEE {
+        return Err(FactoryError::NotEnoughIcp(balance, icp_fee));
     }
 
-    consume_icp(caller, icp_fee, icp_to, ledger).await?;
+    // Transfer the fee to the Factory ID
+    transfer(
+        ledger,
+        icp_fee + TOTAL_FEE,
+        Some((&PrincipalId(caller)).into()),
+        id,
+    )
+    .await?;
+
+    // Tops up the Factory canister with cycles. We only require amount equal to `MIN_CANISTER_CYCLES * 2` + Necessary FEES and send the rest to `icp_to`.
+    let amount = 36_500_000;
```

Comment:
This should be 5 trillion cycles, it should also be written as a constant and stored at the top of the file.